File: solutions/0310. Minimum Height Trees/0310.cpp
// Time complexity: O(V+E)
// Space complexity: O(V+E)
class Solution {
 public:
  vector<int> findMinHeightTrees(int n, const vector<vector<int>>& edges) {
    vector<vector<int>> graph(n);
    vector<int> degree(n, 0);

    for (const vector<int>& edge : edges) {
      const int u = edge[0];
      const int v = edge[1];
      graph[u].push_back(v);
      graph[v].push_back(u);
      degree[u]++;
      degree[v]++;
    }

    vector<int> leaves;
    for (int i = 0; i < n; ++i)
      if (degree[i] <= 1)
        leaves.push_back(i);

    while (n > 2) {
      vector<int> next_leaves;

      for (const int& u : leaves) {
        for (const int& v : graph[u]) {
          degree[v]--;
          if (degree[v] == 1)
            next_leaves.push_back(v);
        }
        n--;
      }
      leaves = next_leaves;
    }

    return leaves;
  }
};

File: solutions/0310. Minimum Height Trees/0310.py
# Time complexity: O(V+E)
# Space complexity: O(V+E)
class Solution:
  def findMinHeightTrees(self, n: int, edges: List[List[int]]) -> List[int]:
    if n == 1 or not edges:
      return [0]
    
    graph = [[] for _ in range(n)]
    degree = [0] * n

    for edge in edges:
      u, v = edge[0], edge[1]
      graph[u].append(v)
      graph[v].append(u)
      degree[u] += 1
      degree[v] += 1

    leaves = [i for i, d in enumerate(degree) if d <= 1]

    while n > 2:
      next_leaves = []

      for u in leaves:
        for v in graph[u]:
          degree[v] -= 1
          if degree[v] == 1:
            next_leaves.append(v)
        n -= 1
      leaves = next_leaves

    return leaves          